contracts/handlers/ERC20Handler.sol

```diff
@@ -57,36 +75,75 @@ contract ERC20Handler is IHandler, ERCHandlerHelpers, ERC20Safe {
+        optionalGas                        uint256 bytes (64 + len(destinationRecipientAddress)) - (96 + len(destinationRecipientAddress))
+        optionalMessage             length uint256 bytes (96 + len(destinationRecipientAddress)) - (128 + len(destinationRecipientAddress))
+        optionalMessage                    bytes   bytes (160 + len(destinationRecipientAddress)) - END
      */
     function executeProposal(bytes32 resourceID, bytes calldata data) external override onlyBridge returns (bytes memory) {
-        uint256       amount;
-        uint256       lenDestinationRecipientAddress;
-        bytes  memory destinationRecipientAddress;
+        uint256 amount;
+        uint256 lenDestinationRecipientAddress;
 
-        (amount, lenDestinationRecipientAddress) = abi.decode(data, (uint, uint));
-        destinationRecipientAddress = bytes(data[64:64 + lenDestinationRecipientAddress]);
+        (amount, lenDestinationRecipientAddress) = abi.decode(data, (uint256, uint256));
+        address recipientAddress = address(bytes20(bytes(data[64:64 + lenDestinationRecipientAddress])));
 
-        bytes20 recipientAddress;
         address tokenAddress = _resourceIDToTokenContractAddress[resourceID];
 
-        assembly {
-            recipientAddress := mload(add(destinationRecipientAddress, 0x20))
+        // Optional message recipient transformation.
+        uint256 pointer = 64 + lenDestinationRecipientAddress;
+        uint256 gas;
+        uint256 messageLength;
+        OptionalMessageCheck optionalMessageCheck;
+        if (data.length > (pointer + 64)) {
+            (gas, messageLength) = abi.decode(data[pointer:], (uint256, uint256));
+            pointer += 64;
+            if (gas > 0 && messageLength > 0 && (messageLength + pointer) <= data.length) {
+                optionalMessageCheck = OptionalMessageCheck.Valid;
+                if (recipientAddress == transformRecipient) {
+                    recipientAddress = _defaultMessageReceiver;
+                }
+            } else {
+                optionalMessageCheck = OptionalMessageCheck.Invalid;
+            }
         }
 
         if (!_tokenContractAddressToTokenProperties[tokenAddress].isWhitelisted) revert ContractAddressNotWhitelisted(tokenAddress);
 
+        uint256 externalAmount = convertToExternalBalance(tokenAddress, amount);
         if (_tokenContractAddressToTokenProperties[tokenAddress].isBurnable) {
-            mintERC20(tokenAddress, address(recipientAddress), convertToExternalBalance(tokenAddress, amount));
+            mintERC20(tokenAddress, recipientAddress, externalAmount);
         } else {
-            releaseERC20(tokenAddress, address(recipientAddress), convertToExternalBalance(tokenAddress, amount));
+            releaseERC20(tokenAddress, recipientAddress, externalAmount);
         }
-        return abi.encode(tokenAddress, address(recipientAddress), amount);
+
+        if (optionalMessageCheck == OptionalMessageCheck.Invalid) {
```

Comment:
Can we please extract OptionalMessage logic into a separate contract from which we extend handlers that should support this logic?
Right now it seems the logic is just duplicated.

Reply:
Added DepositDataHelper.